### PR TITLE
Controllers Branch - Breaking up UserInputMapper, restoring some mappings

### PR DESCRIPTION
**Pull request by jherico**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 734a39f962916e49dfeaa0702a0d13f57d4c1c6d available at: https://gist.github.com/39e8dc426814b7d984b2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6108.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6108.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 184303d3c9a817593052ec64bcc1e04fb8a2a267 available at: https://gist.github.com/39e8dc426814b7d984b2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6108.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6108.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a8e707ced26ed5853d0f0c04dbf24e330d42ce10 available at: https://gist.github.com/39e8dc426814b7d984b2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6108.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6108.dmg

File: libraries/controllers/src/controllers/ScriptingInterface.h [hunk 1]
```diff
@@ -141,6 +165,11 @@ namespace controller {
         ValueMap _overrideValues;
         MappingMap _mappingsByName;
         MappingStack _activeMappings;
+
+        bool _mouseCaptured{ false };
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/ScriptingInterface.h]: the coding standard says `bool _mouseCaptured = false;`

File: libraries/controllers/src/controllers/Pose.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace controller {
+
+    struct Pose {
+    public:
+        vec3 _translation;
+        quat _rotation;
+        vec3 _velocity;
+        quat _angularVelocity;
+        bool _valid{ false };
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/Pose.h]: coding standard says `bool _valid = false;`

File: libraries/controllers/src/controllers/Pose.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+#ifndef hifi_controllers_Pose_h
+#define hifi_controllers_Pose_h
+
+#include <GLMHelpers.h>
+
+namespace controller {
+
+    struct Pose {
+    public:
+        vec3 _translation;
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/Pose.h]: coding standard: PUBLIC member variables should not have leading "_"

File: libraries/controllers/src/controllers/Pose.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+        Pose() {}
+        Pose(const vec3& translation, const quat& rotation,
+             const vec3& velocity = vec3(), const quat& angularVelocity = quat());
+
+        Pose(const Pose&) = default;
+        Pose& operator = (const Pose&) = default;
+        bool operator ==(const Pose& right) const;
+        bool isValid() const { return _valid; }
+        vec3 getTranslation() const { return _translation; }
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/Pose.h]: why provide getters when your members are public? either leave the members as public and drop these getters, or make members private, and provide gettters and setters

File: interface/src/scripting/ControllerScriptingInterface.cpp [hunk 5]
```diff
@@ -183,162 +80,6 @@ const PalmData* ControllerScriptingInterface::getActivePalm(int palmIndex) const
...
-            case PALM_SPATIALCONTROL:
-                return palmData->getPosition();
-            case TIP_SPATIALCONTROL:
-                return palmData->getTipPosition();
-        }
-    }
-    return glm::vec3(0); // bad index
-}
-
-glm::vec3 ControllerScriptingInterface::getSpatialControlVelocity(int controlIndex) const {
```
Comment (ZappoMan) [inline interface/src/scripting/ControllerScriptingInterface.cpp]: The following scripts are broken by this change... we need to either keep this old API, or go into each of these scripts and fix them to use the new technique...
```
  C:\Development\HiFi\hifi\examples\clap.js(74):    var palm1Velocity = Controller.getSpatialControlVelocity(1);
  C:\Development\HiFi\hifi\examples\clap.js(75):    var palm2Velocity = Controller.getSpatialControlVelocity(3);
  C:\Development\HiFi\hifi\examples\controllers\hydra\airGuitar.js(80):        var palmVelocity = Controller.getSpatialControlVelocity(palm * 2 + 1);
  C:\Development\HiFi\hifi\examples\controllers\hydra\drumStick.js(46):        var palmVelocity = Controller.getSpatialControlVelocity(palm * 2 + 1);
  C:\Development\HiFi\hifi\examples\controllers\hydra\frisbee.js(147):    this.tipVelocity = function() { return Controller.getSpatialControlVelocity(this.tip); };
  C:\Development\HiFi\hifi\examples\controllers\hydra\hydraMove.js(172):        grabDeltaVelocity = Controller.getSpatialControlVelocity(RIGHT_PALM);
  C:\Development\HiFi\hifi\examples\controllers\hydra\hydraMove.js(189):        grabDeltaVelocity = Controller.getSpatialControlVelocity(LEFT_PALM);
  C:\Development\HiFi\hifi\examples\controllers\hydra\paddleBall.js(149):                                      velocity: Controller.getSpatialControlVelocity(controllerID),
  C:\Development\HiFi\hifi\examples\controllers\hydra\paddleBall.js(152):                                      velocity: Controller.getSpatialControlVelocity(controllerID),
  C:\Development\HiFi\hifi\examples\controllers\hydra\toyball.js(91):        averageLinearVelocity[0] = Vec3.sum(Vec3.multiply(AVERAGE_FACTOR, Controller.getSpatialControlVelocity(LEFT_TIP)), 
  C:\Development\HiFi\hifi\examples\controllers\hydra\toyball.js(102):        averageLinearVelocity[1] = Vec3.sum(Vec3.multiply(AVERAGE_FACTOR, Controller.getSpatialControlVelocity(RIGHT_TIP)), 
```

Comment (ZappoMan) [inline interface/src/scripting/ControllerScriptingInterface.cpp]: I added a trello card to go and fix all our scripts.

File: libraries/controllers/src/controllers/ScriptingInterface.h (hunk 1 above)
Comment (jherico) [inline libraries/controllers/src/controllers/ScriptingInterface.h]: coding standard should be changed.  email sent.

File: libraries/controllers/src/controllers/Pose.h (hunk 4 above)
Comment (jherico) [inline libraries/controllers/src/controllers/Pose.h]: I don't know man, I didn't do it... I just moved this class out of UserInputMapper.

File: interface/src/scripting/ControllerScriptingInterface.cpp (hunk 5 above)
Comment (jherico) [inline interface/src/scripting/ControllerScriptingInterface.cpp]: I can go either way...  I think the right thing to do is update the scripts, but I do have concerns about breaking user scripts we don't control.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 22602fb6f6506c5af6f7fa04ca51573c41d393bd available at: https://gist.github.com/39e8dc426814b7d984b2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6108.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6108.dmg

**Pull request merged**